overlord/snapstate/refresh.go

```diff
@@ -184,3 +157,87 @@ func (err *BusySnapError) Error() string {
 func (err BusySnapError) Pids() []int {
 	return err.pids
 }
+
+// doHardRefreshCheck performs the complete hard refresh check interaction.
+//
+// This check embeds HardNothingRunningRefreshCheck along with interaction
+// with two locks - the snap lock, shared by snap-confine and snapd and the
+// snap run inhibition lock, shared by snapd and snap run.
+//
+// On success this function returns a locked lock, allowing the caller to
+// atomically, with regards to snap-confine, finish any action that required
+// the check.
+//
```

Comment:
this comments doesn't explain clearly what is done with the "snap run inhibition lock" ...

Also in hindsight we have two inhibitions that are complementary:

* we either inhibit the refresh
* or we inhibit the running and do the refresh

that results in use needing to be careful with the term inhibit alone, or we need to rename those concepts

Reply:
Good point. I re-wrote this whole paragraph.